presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java

```diff
@@ -229,15 +244,20 @@ public NodeSelector createNodeSelector(ConnectorId connectorId, int maxTasksPerS
         return new ResettableRandomizedIterator<>(nodes);
     }
 
-    public static List<InternalNode> selectExactNodes(NodeMap nodeMap, List<HostAddress> hosts, boolean includeCoordinator)
+    public static List<Pair<NodeStatus, InternalNode>> selectExactNodes(NodeMap nodeMap, List<HostAddress> hosts, boolean includeCoordinator)
     {
-        Set<InternalNode> chosen = new LinkedHashSet<>();
+        Set<Pair<NodeStatus, InternalNode>> chosen = new LinkedHashSet<>();
```

Comment:
Can `NodeStatus` be made as a field of `InternalNode`?

Reply:
The `List<InternalNode>` returned by nodeManager are always alive. If a node is dead nodemanger doesnt even return  it. Also InternalNode is thrift struct. If we move it into InternalNode, we have to compare the node list returned by nodemanager with cached node list and figure out which nodes are dead and then set nodeStatus=DEAD for those InternalNode. I think having nodeStatus out is more cleaner.

Reply:
Thrift struct is a good point. I'm not aware of that as I did not see thrift annotations in the class.

For the second reason, is it possible to handle it in `DiscoveryNodeManager`? This might already be explored but I'm basically looking at how missing active nodes and failed nodes are being handled in: 
https://github.com/prestodb/presto/blob/3692a11a7bbab443dfe8d731ec1347a06d6d09b3/presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java#L227

Maybe we could extend it a bit?

Reply:
I think this was discussed before. The discovery service Inactive nodes are completely different, and probably is a bad name.  Inactive nodes just means version mismatch nodes.

Reply:
yeah. If a node is down, it will not even show up in the `for loop` you pointed above. INACTIVE nodes are alive workers, but just that their version doesnt match with that of coordinator.

Reply:
Yeah I might be wrong. But I kinda feel The difference between `allNodes` (not refreshed yet) and `services` is the failed nodes. What am I missing? 🤔

Reply:
good point ! We can get the dead nodes from here. And i see we already log that. May be we can tweak that.
https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java#L297